File: Difficulty/Medium/[17] Letter Combinations of a Phone Number/solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res, temp;
        if (digits.size() == 0)
            return res;
        int s = 0, n = 0, b = 0;
        string ts;
        for (int i = 0; i < digits.size(); ++i){
            s = int(digits[i]) - 48;
            n = 3 + (abs(s - 8) == 1);
            b = 97 + 3 * (s - 2) + (s > 7);
            if (i == 0)
                for (int j = 0; j < n; ++j){
                    ts = char(b + j);
                    res.push_back(ts);
                }
            else{
                for (int j = 0; j < res.size(); ++j){
                    for (int k = 0; k < n; ++k)
                        temp.push_back(res[j] + char(b + k));
                }
                res = temp;
            }
            temp.clear();
        }
        return res;
    }
};

File: Difficulty/Medium/[17] Letter Combinations of a Phone Number/solution.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty()) return new ArrayList<>();
        List<String> res = new ArrayList<>();
        HashMap<Character, String> map = new HashMap<>() {
            {
                put('2', "abc");
                put('3', "def");
                put('4', "ghi");
                put('5', "jkl");
                put('6', "mno");
                put('7', "pqrs");
                put('8', "tuv");
                put('9', "wxyz");
            }
        };
        StringBuilder sBuf = new StringBuilder(digits.length());
        dfs(map, digits, res, sBuf, 0);
        return res;
    }

    public void dfs(HashMap<Character, String> map, String digits, List<String> res, StringBuilder sBuf, int i) {
        if (i == digits.length()) {
            res.add(sBuf.toString());
            return;
        }
        char c = digits.charAt(i);
        int len = map.get(c).length();
        for (int j = 0; j < len; j++) {
            sBuf.append(map.get(c).charAt(j));
            dfs(map, digits, res, sBuf, i+1);
            sBuf.deleteCharAt(sBuf.length() - 1);
        }
    }
}